src/System.Text.Primitives/System/Text/Encoders/Utf8Decoder.cs

```diff
@@ -0,0 +1,264 @@
+                return DangerousReadUnicodeScalarValueWithoutNullCheck(ref data, bytesAvailable, out bytesConsumed);
+            }
+        }
+
+        // Assumption: bytesAvailable > 0
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int DangerousReadUnicodeScalarValueWithoutNullCheck(ref byte data, int bytesAvailable, out int bytesConsumed)
+        {
+            Debug.Assert(bytesAvailable > 0);
+
+            if ((data & 0x80) == 0)
+            {
+                // Fast case: ASCII character => the code point is the byte itself
+                bytesConsumed = 1;
+                return data;
+            }
+            else
+            {
+                // Slow case: multi-byte sequence
+                bytesConsumed = 0;
+                return ReadUnicodeScalarValueFromNonAscii(ref data, bytesAvailable, ref bytesConsumed);
+            }
+        }
+
+        /// <summary>
+        /// Retrieves the index of the first invalid byte in the data, i.e., the index of the
+        /// first invalid UTF8 subsequence within the data.
+        /// </summary>
+        /// <param name="data">The data to check.</param>
+        /// <returns>The index into <paramref name="data"/> where the first invalid byte appears, or -1
+        /// if <paramref name="data"/> represents a well-formed UTF8 string.</returns>
+        /// <remarks>An invalid subsequence could occur because of the presence of an illegal byte
+        /// (such as 0xFF) or because <paramref name="data"/> begins or ends in the middle of a subsequence.</remarks>
+        public static int GetIndexOfFirstInvalidUtf8Byte(ReadOnlySpan<byte> data)
+        {
+            // TODO: Can this be optimized, perhaps by performing a clever "is ASCII?" bitwise check
+            // and consuming entire ASCII sequences one natural word length at a time?
+
+            int numValidBytesSoFar = 0;
+
+            while (numValidBytesSoFar < data.Length)
+            {
+                if (DangerousReadUnicodeScalarValueWithoutNullCheck(ref Unsafe.Add(ref data.DangerousGetPinnableReference(), numValidBytesSoFar), data.Length - numValidBytesSoFar, out int thisIterBytesConsumed) < 0)
+                {
+                    return numValidBytesSoFar; // error (not enough data or malformed sequence seen)
+                }
+                else
+                {
+                    numValidBytesSoFar += thisIterBytesConsumed; // success (read a valid code point from the sequence)
+                }
+            }
+
+            return -1; // no invalid bytes
+        }
+
+        /// <summary>
+        /// Given the first byte of a UTF8 multi-byte sequence, returns the number of expected trailing
+        /// bytes and the minimum allowed code point value which is encoded by this sequence. Returns
+        /// 0 on invalid leading byte.
+        /// </summary>
+        private static int GetNumberOfTrailingBytes(byte firstByte, out int minAllowedCodePointValue)
```

Comment:
Would any of the helpers in here be a helpful substitute for this, particularly:
https://github.com/dotnet/corefxlab/blob/master/src/System.Text.Primitives/System/Text/Encoders/EncodingHelper.cs#L116?

If not, maybe we should add this method there.